File: 03-march/20-minimum-absolute-difference-in-sliding-submatrix/solution.py
class Solution:
    def minAbsDiff(self, grid: list[list[int]], k: int) -> list[list[int]]:
        rows = len(grid)
        cols = len(grid[0])
        ans = []
        for i in range(rows - k + 1):
            row_ans = []
            for j in range(cols - k + 1):
                values = []
                seen = set()
                for x in range(i, i + k):
                    for y in range(j, j + k):
                        val = grid[x][y]
                        if val not in seen:
                            values.append(val)
                            seen.add(val)
                if len(values) <= 1:
                    row_ans.append(0)
                else:
                    values.sort()
                    diff = min(values[p + 1] - values[p] for p in range(len(values) - 1))
                    row_ans.append(diff)
            ans.append(row_ans)
        return ans

File: 03-march/20-minimum-absolute-difference-in-sliding-submatrix/solution.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 */
var minAbsDiff = function(grid, k) {
  const rows = grid.length;
  const cols = grid[0].length;
  const ans = [];

  for (let i = 0; i <= rows - k; i++) {
    const rowAns = [];
    for (let j = 0; j <= cols - k; j++) {
      const values = [];
      const seen = new Set();
      for (let x = i; x < i + k; x++) {
        for (let y = j; y < j + k; y++) {
          const val = grid[x][y];
          if (!seen.has(val)) {
            seen.add(val);
            values.push(val);
          }
        }
      }
      if (values.length <= 1) {
        rowAns.push(0);
      } else {
        values.sort((a, b) => a - b);
        let diff = Number.MAX_SAFE_INTEGER;
        for (let m = 0; m < values.length - 1; m++) {
          diff = Math.min(diff, values[m + 1] - values[m]);
        }
        rowAns.push(diff);
      }
    }
    ans.push(rowAns);
  }

  return ans;
};
